### PR TITLE
Introduce staged timeout system

### PR DESCRIPTION
**Pull request by M-Davies**
All timeouts are now done on a stage by stage basis, failing the build with a helpful message if the timeout is reached:

**TOP LEVEL PIPELINES**
* Adopt api requests = 1 hour
* Copy artifacts = 6 hours
* Remove artifacts = 2 hours
* Archive artifacts = 6 hours
* Publish artifacts = 3 hours

**DOWNSTREAM JOBS**
* Clean master node = 1 hour
* Clean aix node = 1 hour
* Adopt api requests = 1 hour
* Clean workspace on node = 1 hour
* Checkout scm on node = 1 hour
* Pull docker image = 2 hours
* Checkout docker image = 1 hour
* Main build script (make-adopt-build-farm.sh) = 6 hours
* Archive artifact = 3 hours
* Sign build job = 2 hours
* Installer jobs = 3 hours

- I have added the nessasary stubs and libraries to ensure that the try-catch fires correctly when a timeout is reached. For the FlowInterruptedException, it required importing the jenkins core lib via gradle (https://stackoverflow.com/a/47353745/14420589).

- While I was at it, I also added in a node timeout to the pr tester (since that was a param that got implemented recently https://github.com/AdoptOpenJDK/openjdk-build/issues/2118) as well as correcting the function that ran it to use Jenkins advanced sleep which has more brevity and reliability than normal groovy sleep. I also added a println to when we move into/out-of nodes as it's hard to tell sometimes in the logs

Closes: #2120 
Signed-off-by: Morgan Davies <morgandavies2020@gmail.com>

### CODE REVIEW

**Comment by M-Davies** [comment]
run tests

File: pipelines/build/common/build_base_file.groovy
Comment (karianna) [inline pipelines/build/common/build_base_file.groovy:334]: I'd parameterise these error messages with the timeout logic above (as opposed to writing in a string)

**Review by karianna — CHANGES_REQUESTED** [review]

File: pipelines/build/common/build_base_file.groovy
Comment (M-Davies) [reply pipelines/build/common/build_base_file.groovy:334]: That's the plan! I'm still working on getting the gradle dependancies sorted out first, then I'm going to look into the output messages

**Comment by M-Davies** [comment]
[Latest tester](https://ci.adoptopenjdk.net/job/build-scripts-pr-tester/job/openjdk-build-pr-tester/1123/console) failed due to [ongoing aarch64 build issues](https://adoptopenjdk.slack.com/archives/C09NW3L2J/p1603096291133100). Otherwise, all builds passed without timing out or failing the gradle download step. I'll param the error messages, squash and edit the commits and move this into review ready

File: pipelines/build/common/build_base_file.groovy
Comment (karianna) [inline pipelines/build/common/build_base_file.groovy:442]: This is quite a complex construct now.  Consider extracting logic to local private functions that can be called from within this block

**Review by karianna — CHANGES_REQUESTED** [review]

File: pipelines/build/common/build_base_file.groovy
Comment (M-Davies) [reply pipelines/build/common/build_base_file.groovy:442]: Short of rebasing the entire function or passing in half of the params into these private functions, there's not much we can really do here to make it simpler. If we do decide to go ahead with that, I feel that should be done in a separate issue as that would be a big job.

Comment (karianna) [reply pipelines/build/common/build_base_file.groovy:442]: @johnoliver Ay insights here from your experience of pyramids of doom with Java?

**Comment by M-Davies** [comment]
run tests

File: pipelines/build/common/build_base_file.groovy
Comment (johnoliver) [reply pipelines/build/common/build_base_file.groovy:524]: I think the only thing you can do is extract the functions and pass the mass of parameters, not much way around it

**Review by karianna — APPROVED** [review]

**Review by andrew-m-leonard — APPROVED** [review]
lgtm, needs a good test though

**Comment by M-Davies** [comment]
> lgtm, needs a good test though

Agreed although I'm wondering what the best way to test it is. In the meantime (seeing as people seem happy with the timeout values), I'll put this back into draft and figure out a way to test it

**Comment by M-Davies** [comment]
@andrew-m-leonard @karianna Would we prefer to have the timeouts defined on a platform to version basis instead of a general timeout for every platform and version? I'm thinking that we could add in a Map attribute into each platform's [build_configuration_map](https://github.com/AdoptOpenJDK/openjdk-build/blob/master/pipelines/jobs/configurations/jdk8u_pipeline_config.groovy) that contains the timeouts for that specific platform. That way we could have longer timeouts for slower systems like AIX and ARM and shorter ones for others?

File: pipelines/build/common/openjdk_build_pipeline.groovy
Comment (AdamBrousseau) [inline pipelines/build/common/openjdk_build_pipeline.groovy:74]: Since I don't see any test team members on this PR yet I suggest we run this by @smlambert and @llxia. I know the test pipelines have their own, configurable timeout value. If we didn't have one for test would it cause a timeout "blindspot" or would it get covered in a higher level timeout?

Comment (smlambert) [reply pipelines/build/common/openjdk_build_pipeline.groovy:74]: Thanks @AdamBrousseau - we do in fact manage our own timeouts in the test pipelines, because philosophically we believe each stage and step in the pipeline should manage their own timeouts and as there are valid cases where the default of 10hrs is different.

If the build pipelines want to enforce a 10hr limit, this PR should pass TEST_JOBS_TIMEOUT as a parameter in [openjdk_build_pipeline.groovy L224-L230](https://github.com/AdoptOpenJDK/openjdk-build/blob/master/pipelines/build/common/openjdk_build_pipeline.groovy#L224-L230).  The parameter name to set is called [TIME_LIMIT](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/buildenv/jenkins/JenkinsfileBase#L660).  

Centrally managing and hardcoding all of these timeouts from a build pipeline makes this build pipeline code very closely tied to the performance of the Adopt build/test farm.  I would expect a _future request_, if we are building flexible/portable pipeline code will be to allow for all of these settings to be tune-able (without having to maintain a fork of this repo).

Comment (M-Davies) [reply pipelines/build/common/openjdk_build_pipeline.groovy:74]: If the test pipelines have their own timeouts, I’m leaning towards just removing the timeouts I want to introduce from openjdk_build_pipeline.groovy completely. I don’t really want to have a clash of timeouts that counter each other

**Comment by M-Davies** [comment]
run tests

**Comment by M-Davies** [comment]
Tester failed due to https://github.com/AdoptOpenJDK/openjdk-build/pull/2212. This is blocked until that PR is merged

**Comment by karianna** [comment]
Conflicts as well

**Comment by M-Davies** [comment]
> Conflicts as well

I'm waiting till 2212 is merged before resolving them since it means I only have to do it once 🙂

**Comment by M-Davies** [comment]
run tests

**Comment by andrew-m-leonard** [comment]
> @andrew-m-leonard @karianna Would we prefer to have the timeouts defined on a platform to version basis instead of a general timeout for every platform and version? I'm thinking that we could add in a Map attribute into each platform's [build_configuration_map](https://github.com/AdoptOpenJDK/openjdk-build/blob/master/pipelines/jobs/configurations/jdk8u_pipeline_config.groovy) that contains the timeouts for that specific platform. That way we could have longer timeouts for slower systems like AIX and ARM and shorter ones for others?

I'm wary of over complicating things here, and also for example a particular AIX machine maybe slow now, but a simple H/W upgrade could easily change that. I would stick with a "course" set of timeouts that roughly covers the given stage.

**Comment by M-Davies** [comment]
Thanks Andrew. In that case, I'll leave as it is for now and let others add in platform specific timeouts in the future should they prove necessary. This is ready for review again (and merge, pending PR tester result)

**Comment by M-Davies** [comment]
run tests

**Comment by M-Davies** [comment]
run tests

**Comment by M-Davies** [comment]
run tests

**Comment by M-Davies** [comment]
run tests

**Comment by M-Davies** [comment]
> # 🟠 PR TESTER RESULT 🟠
> ❎ Some pipelines failed or the job was aborted! ❎
> _See the `pipeline-build-check` below for more information..._

Windows failed due to https://github.com/AdoptOpenJDK/openjdk-infrastructure/issues/1573
[aarch64/JDK16/j9](https://ci.adoptopenjdk.net/job/build-scripts-pr-tester/job/build-test/job/jobs/job/jdk/job/jdk-linux-aarch64-openj9/382/console) failed due to unknown error. Rebuild passed https://ci.adoptopenjdk.net/job/build-scripts-pr-tester/job/build-test/job/jobs/job/jdk/job/jdk-linux-aarch64-openj9/384/
```
13:02:34  CCACHE_COMPRESS=1          CCACHE_SLOPPINESS=pch_defines,time_macros CCACHE_BASEDIR=/home/ubuntu/workspace/build-scripts-pr-tester/build-test/jobs/jdk/jdk-linux-aarch64-openj9/workspace/build/src /usr/local/gcc/bin/ccache  /usr/local/gcc/bin/gcc-7.5 -O -fgnu89-inline -g -DLINUX -D_REENTRANT -D_FILE_OFFSET_BITS=64 -fpic -DIPv6_FUNCTION_SUPPORT -DJ9AARCH64 -fstack-protector -Wimplicit -Wreturn-type -Werror -I. -I../include -I../oti -I../gc_include -I../omr/gc/include -I../gc_glue_java -I../nls -I../omr/include_core    -DTR_HOST_ARM64   -E vmcheck.c | sed -n -e '/^DDRFILE_BEGIN /,/^DDRFILE_END /s/^/@/' -e '/^@./p' > vmcheck.i
13:02:35  gmake[6]: Leaving directory '/home/ubuntu/workspace/build-scripts-pr-tester/build-test/jobs/jdk/jdk-linux-aarch64-openj9/workspace/build/src/build/linux-aarch64-server-release/vm/vmchk'
13:02:35  gmake[5]: Leaving directory '/home/ubuntu/workspace/build-scripts-pr-tester/build-test/jobs/jdk/jdk-linux-aarch64-openj9/workspace/build/src/build/linux-aarch64-server-release/vm'
13:02:35  [2020-11-09 13:02:34] Scraping anotations from preprocessed code ...
13:02:41  [2020-11-09 13:02:40] Restoring annotated files ...
13:02:41  [2020-11-09 13:02:41] All done.
13:02:41  Running ddrgen to generate j9ddr.dat and superset.dat
13:03:54  Blob written to file: ../j9ddr.dat
13:03:54  Superset written to file: ../superset.dat
13:03:54  gmake[4]: Leaving directory '/home/ubuntu/workspace/build-scripts-pr-tester/build-test/jobs/jdk/jdk-linux-aarch64-openj9/workspace/build/src/build/linux-aarch64-server-release/vm/ddr'
13:03:54  
13:03:54  ERROR: Build failed for targets 'product-images legacy-jre-image test-image debug-image' in configuration 'linux-aarch64-server-release' (exit code 2) 
13:03:54  Stopping sjavac server
13:03:54  
13:03:54  === Output from failing command(s) repeated here ===
13:03:54  * For target jdk_modules_jdk.jshell__the.jdk.jshell_batch:
13:03:54  * For target jdk_modules_jdk.security.auth__the.jdk.security.auth_batch:
13:03:54  
13:03:54  * All command lines available in /home/ubuntu/workspace/build-scripts-pr-tester/build-test/jobs/jdk/jdk-linux-aarch64-openj9/workspace/build/src/build/linux-aarch64-server-release/make-support/failure-logs.
13:03:54  === End of repeated output ===
13:03:54  
13:03:54  === Make failed targets repeated here ===
13:03:54  CompileJavaModules.gmk:609: recipe for target '/home/ubuntu/workspace/build-scripts-pr-tester/build-test/jobs/jdk/jdk-linux-aarch64-openj9/workspace/build/src/build/linux-aarch64-server-release/jdk/modules/jdk.security.auth/_the.jdk.security.auth_batch' failed
13:03:54  make/Main.gmk:197: recipe for target 'jdk.security.auth-java' failed
13:03:54  CompileJavaModules.gmk:609: recipe for target '/home/ubuntu/workspace/build-scripts-pr-tester/build-test/jobs/jdk/jdk-linux-aarch64-openj9/workspace/build/src/build/linux-aarch64-server-release/jdk/modules/jdk.jshell/_the.jdk.jshell_batch' failed
13:03:54  make/Main.gmk:197: recipe for target 'jdk.jshell-java' failed
13:03:54  === End of repeated output ===
```

**Comment by M-Davies** [comment]
@karianna Assuming @smlambert is happy with these changes, this should be good to merge

**Review by smlambert — APPROVED** [review]
thanks @M-Davies !

**Pull request merged**
